[cire] Step-2 GET 요청으로 회원가입 기능 구현

cire0304 opened the pull request:
# 아쉬운 점 1

예외 처리에 대한 부분이 너무 아쉽다.
나중에 시간이 난다면, 따로 구현을 해야할 것 같다.

# 아쉬운 점 2

요청 파라미터를 객체로 변환해주는 ObjectMapper를 따라한 UserMapper를 구현했지만, 너무 이상하게 구현을 했다..

1. 먼저 객체의 생성자 매개변수와, 요청 파라미터의 순서가 맞아야 한다는 문제가 있다. (나의 수준으론 해결이 어려울 것 같다 ㅜㅜ)
2. 리플렉트 API를 사용하면, OOP에 위배될 수 있어 사용하면 좋지 않다고 하는데, 사용했다. 
-> 왜 사용하면 안되는지 모르면서 그냥 사용 (생성자의 매개변수만 참고하는 정도는 될 것같은데 말이다)
3. 예외처리도 제대로 되지 않았다. (나중에..)

# 좋았던 점

그래도 이상하게 구현하긴 했지만, 배운점도 있었다.

스프링의 코드를 보면 가끔씩 `Class<?>` 가 왜 나오는지 몰랐지만, 이번에 알게됬다.

cire0304 committed:
Style : 코드의 가독성을 위한 메소드 분리 적용

기존의 코드는 메소드를 분리 하지 앟아 가독성이 많이 떨어졌다.

cire0304 committed:
Rename : 명시적인 클래스 이름으로 변경

cire0304 committed:
Refactory : RequestProcess 클래스의 상태 제거

맴버 변수 HttpReqestHead를 제거하였다.

현재의 코드에서 문제는 없지만, 쓰레드풀을 사용하여 구현하게 된다면,
이전의 요청의RequestProcess 상태가 이후 요청에 영향을 끼칠 가능성이 생길 것을 방지하기 위해  맴버 변수 HttpReqestHead를 제거하였다.

cire0304 committed:
Refactor : 책임 분리를 위한 코드 수정

URL에 대한 정보에 대한 처리는 HttpReqeustHead 에 대한 책임이다

cire0304 committed:
Rename : 메소드를 명시적인 이름으로 수정

cire0304 committed:
Refacotr : 테스트 실행 시간 단축을 위한 코드 변경

하나의 head 스트링으로 모든 테스트를 진행

cire0304 committed:
Style : 코드의 가독성을 위한 메소드 분리 적용

cire0304 committed:
Rename : 가독성을 위한 변수 이름 변경

cire0304 committed:
Rename : 정확한 용어 사용을 위한 이름 변경

cire0304 committed:
Style : 코드의 가독성을 위한 메소드 분리 적용

cire0304 committed:
Test : 테스트 코드 분리 및 테스트 추가

기존 테스트 코드는 한번의 테스트로 여러가지 결과를 테스트 하려고 했다.
그러므로 테스트 코드를 분리하였다.

또한, Display이름을 변경하였다.

cire0304 committed:
Feat : 요청 파라미터 관련 기능 추가

요청 파라미터를 파싱하기 위해, 메소드를 추가하였다.

cire0304 committed:
Deprecate :  공부용 코드

cire0304 committed:
Feat : 요청 파라미터를 User 객체로 변환하는 기능 구현

cire0304 committed:
Refactor : 책임 분리를 위한 코드 변경

HttpRequestHead 의 객체는, 단순히 HTTP 요청의 정보를 가지고 있는 책임만 가지고 있어야할 것같다.

정보를 처리 및 변환하는 책임은 다른얘한테 줘보자 (시도)

cire0304 committed:
Test : UserMapper 테스트 코드 작성

cire0304 committed:
Remove : 불필요한 테스트 코드 삭제

honux77 commented on src/test/java/process/RequestProcessTest.java:
```diff
@@ -24,15 +30,33 @@ public void beforeEach() {
         RequestProcess requestProcess = new RequestProcess();
 
-        byte[] body = requestProcess.process(head);
+        byte[] body = requestProcess.makeResponse(head);
```
head 스트링으로 한번에 body를 받아버리는군요. 
너무 큰일을 하는 것 같은데 조금 쪼개도 좋을 것 같아요.

honux77 commented on src/main/java/model/HttpRequestHead.java:
```diff
@@ -0,0 +1,70 @@
+    private String url;
+    private String requestParams;
+    private final String version;
+    private final Map<String, String> MIMEHeader;
```
MIME이 여기서 왜 나오는지 모르겠어요.

honux77 replied:
startLine을 제외한 나머지를 :으로 구분해서 key value로 나누려고 한 것 같은데 그것과 MIME은 다릅니다. 적절한 변수명과 메소드명을 생각해서 고쳐야 할 것 같아요. 변수명 때문에 제가 엄청 오랫동안 고민을 했어요 ㅜㅜ

cire0304 replied:
죄송합니다 ㅜㅜ
표준 HTTP 헤더 형식을 좀 더 공부하겠습니다!!

honux77 reviewed (changes requested):
HTTP Request에 대해 공부를 하고 메소드 이름을 바꿔주세요.

cire0304 committed:
Refactor : UserMapper 코드 변경

private로 선언된 맴버 변수를 이용하여, 생성자를 찾는 것은 OOP 원칙에 위배된다.
따라서 public으로 선언된 생성자를 이용하여, 객체를 만들도록 코드 변경

cire0304 committed:
Feat : UserMapper를 MyObjectMapper로 변경

범용적으로 Mapper 기능을 사용하기 위해 코드 변경

cire0304 committed:
Fix : 버그 수정

잘못된 연산을 하고 있었다.

cire0304 committed:
Rename : 잘못된 용어 사용 변경

cire0304 committed:
Remove : 사용되지 않는 코드 삭제

cire0304 committed:
Feat : HTTP 요청 메소드 ENUM 타입 추가

요청 메소드를 타입으로 구분하기 위해 ENUM 타입 추가

cire0304 committed:
Refactor : GET 요청에 대한 책임 분리

죄송합니다... 스프링을 따라하는 것 말곤 책임을 분리 못하겠습니다 ㅜㅜ

cire0304 committed:
Test : 테스트 코드 변경

cire0304 committed:
Remove : 사용되지 않는 코드 삭제

cire0304 committed:
Test : 유저 회원 가입 기능 테스트 코드 작성

honux77 commented on src/main/java/webserver/RequestHandler.java:
```diff
@@ -25,25 +27,46 @@ public void run() {
+    private static String readRequestHead(BufferedReader br) throws IOException {
+        StringBuilder stringBuilder = new StringBuilder();
+        readRequestLine(br, stringBuilder);
+        readRequestMIME(br, stringBuilder);
```
여전히 여기에도 남아 있네요. 이것도 고쳐주세요.

honux77 commented on src/main/java/webserver/RequestHandler.java:
```diff
@@ -25,25 +27,46 @@ public void run() {
-                stringBuilder.append(line).append("\n");
-            }
-            stringBuilder.append("\n");
+            String requestHead = readRequestHead(br);
```
```suggestion
            List<String> requestHead = readRequestHead(br);            
```
\n 으로 구분하는 것보다 이게 좋지 않을까요?

honux77 reviewed (changes requested):
여전히 MIME 이름의 메소드가 남아 있는데 왜 사용했는지 궁금합니다. 오후에 잠시 이야기 해봐요.

cire0304 committed:
Rename : 잘못된 용어를 사용한 이름 변경

honux77 commented on src/test/java/service/UserServiceTest.java:
```diff
@@ -0,0 +1,34 @@
+
+    }
+
+}
```
간만에 EOL ㅎㅎ

cire0304 committed:
Chore : EOL 추가

honux77 commented on src/main/java/util/MyObjectMapper.java:
```diff
@@ -0,0 +1,57 @@
+        List<String> params = parseRequestParams(requestParams);
+
+        // TODO : 현재의 객체의 생성자를 찾는 방법은 파라미터의 개수만을 이용하여 찾는다.
+        // TODO : 생성자 매개변수의 이름도 사용할 수 없을까?
```
뭐든 가능합니다 ^^

honux77 commented on src/main/java/util/MyObjectMapper.java:
```diff
@@ -0,0 +1,57 @@
+import java.lang.reflect.InvocationTargetException;
+import java.util.*;
+
+public class MyObjectMapper {
```
이렇게 시도해 본 것도 좋은 것 같아요.

honux77 reviewed (approved):
시레 고생하셨습니다. 화이팅!
